### PR TITLE
add support for postgresql driver in function sdk

### PR DESCRIPTION
**Pull request by mangalaman93**
Completes HYP-1323
Also completes HYP-1324

### CODE REVIEW

File: src/assembly/datasource.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class pgResponse {
+  result!: string;
+  error!: string;
+}
+
+export abstract class postgresql {
+  static query(
+    datasourceName: string,
+    query: string,
+    params: string[],
```
Comment (mattjohnsonpint) [inline src/assembly/datasource.ts:22]: Params can be optional.
```suggestion
    params: string[] = [],
```

File: examples/postgresql/assembly/index.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { postgresql } from "@hypermode/functions-as";
+
+export function queryPeople(): string {
+  return postgresql.query("neon", "select * from people", []);
```
Comment (mattjohnsonpint) [inline examples/postgresql/assembly/index.ts:4]:
```suggestion
  return postgresql.query("neon", "select * from people");
```

File: examples/postgresql/assembly/index.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  postgresql.query("neon", query, [
+    name,
+    age.toString(),
+    balance.toString(),
+    privkey.toString(),
+    new Date(Date.now()).toUTCString(),
+    male.toString(),
+    "(" + lat.toString() + "," + lon.toString() + ")",
+  ]);
```
Comment (mattjohnsonpint) [inline examples/postgresql/assembly/index.ts:33]: This isn't great.  The user shouldn't have to turn everything into a string.

It also means that we're passing all parameters as strings into Postgres, such as:

```json
["John","42,123.45","[1,2,3,4]","2024-06-19T23:59:59.999Z","true","(123.45,-67.89)"]
```

The existing `QueryVariables` class could be re-purposed here, which would turn this code into something like:

```ts
  const params = new QueryVariables();
  params.set("$1", name);
  params.set("$2", age);
  params.set("$3", balance);
  params.set("$4", privkey);
  params.set("$5", new Date(Date.now()));
  params.set("$6", male);
  params.set("$7", `(${lat},${lon})`);

  postgresql.query("neon", query, params);
```

However, since postgres only use positional parameters (unlike SQL Server and others that use named parameters), we should make a new class in our library:

```ts
export class QueryParameters {
  private data: string[] = [];

  public add<T>(value: T): void {
    this.data.push(JSON.stringify(value));
  }

  public toJSON(): string {
    return `[${this.data.join(",")}]`;
  }
}
```

Then the usage can be:

```ts
  const params = new QueryParameters();
  params.add(name);
  params.add(age);
  params.add(balance);
  params.add(privkey);
  params.add(new Date(Date.now()));
  params.add(male);
  params.add(`(${lat},${lon})`);

  postgresql.query("neon", query, params);
```

The resulting JSON array would then not contain only strings, but for example:

```json
["John",42,123.45,[1,2,3,4],"2024-06-19T23:59:59.999Z",true,"(123.45,-67.89)"]
```

Comment (mattjohnsonpint) [reply examples/postgresql/assembly/index.ts:33]: Also you could try passing the lat/lon as an array - I'm not sure how Postges needs it though.

```ts
params.add(<f32[]>[lat, lon]);
```

File: examples/postgresql/assembly/index.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+export function getPerson(name: string): string {
+  const query = "select * from people where name = $1";
+  return postgresql.query("neon", query, [name]);
+}
+
+export function addPerson(
+  name: string,
+  age: i32,
+  balance: f64,
+  privkey: i8[],
```
Comment (mattjohnsonpint) [inline examples/postgresql/assembly/index.ts:16]:
```suggestion
  privkey: u8[],
```

Or:

```suggestion
  privkey: Uint8Array,
```

Or:
```suggestion
  privkey: ArrayBuffer,
```

Ideally the runtime should support all three (we should test that), but either way it would be unsigned.

File: src/assembly/datasource.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    const input: pgInput = {
+      query: query,
+      params: params,
+    };
+
+    const respStr = host.databaseQuery(
+      datasourceName,
+      datasourceType,
+      JSON.stringify<pgInput>(input),
+    );
```
Comment (mattjohnsonpint) [inline src/assembly/datasource.ts:35]: Lets pass query and params as two separate variable to the host function.  That will reduce some JSON work for the query, and will make it easier to use the `QueryParameters` object mentioned in the comment in the example.

```suggestion
    const respStr = host.databaseQuery(
      datasourceName,
      datasourceType,
      query,
      params.toJSON()
    );
```

**Review by mattjohnsonpint — CHANGES_REQUESTED** [review]

File: src/assembly/datasource.ts (hunk 1 above)
Comment (mattjohnsonpint) [reply src/assembly/datasource.ts:22]: ... or preferably if you implement `QueryParameters` as mentioned in my feedback on the example, then it would be:

```suggestion
    params: QueryParameters | null = null,
```

File: examples/postgresql/assembly/index.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import { postgresql } from "@hypermode/functions-as";
+
+export function queryPeople(): string {
+  return postgresql.query("neon", "select * from people", []);
+}
+
+export function getPerson(name: string): string {
+  const query = "select * from people where name = $1";
+  return postgresql.query("neon", query, [name]);
+}
```
Comment (mattjohnsonpint) [inline examples/postgresql/assembly/index.ts:10]: This example is to get a _person_, but instead we get a `string` - which the user would have to decide how to parse.  Since we're the ones in control over how the API returns data, we should return it already parsed to the format that the user is expecting.  In other words, the user should not have to think too much about JSON to interpret the results.

To achieve that, the user can pass in an object to be populated as a generic type parameter.  This pattern is already being used on the GraphQL example.

That also gives the user an opportunity to see how to use the results in their own application logic.

In other words:

```suggestion
export function getPerson(name: string): Person {
  const query = "select * from people where name = $1";
  const people = postgresql.query<Person>("neon", query, [name]);
  if (people.length) == 0 {
    throw new Error(`Person not found named ${name}.`);
  }
  if (people.length > 1) {
    throw new Error(`More than one person found named ${name}.`);
  }
  
  return people[0];
}
```

File: examples/postgresql/assembly/index.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { postgresql } from "@hypermode/functions-as";
+
+export function queryPeople(): string {
+  return postgresql.query("neon", "select * from people", []);
+}
```
Comment (mattjohnsonpint) [inline examples/postgresql/assembly/index.ts:5]:
```suggestion
export function queryPeople(): Person[] {
  return postgresql.query<Person>("neon", "select * from people");
}
```

(see feedback on `getPerson` example)

File: src/assembly/datasource.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  result!: string;
+  error!: string;
+}
+
+export abstract class postgresql {
+  static query(
+    datasourceName: string,
+    query: string,
+    params: string[],
+  ): string {
```
Comment (mattjohnsonpint) [inline src/assembly/datasource.ts:23]:
```suggestion
  static query<TData>(
    datasourceName: string,
    query: string,
    params: QueryParameters | null = null,
  ): TData[] {
```

see feedback on the examples.

**Comment by mattjohnsonpint** [comment]
Updated by #119  (which includes much of this)
Thanks.
